### PR TITLE
free_symbols and find_dynamicsymbols methods for vectors directly 

### LINKED ISSUE #13398: physics.vector.Vector should have atoms and free_symbols
It would be nice to be able to call `free_symbols` on a physics vector and it works as it does with any expression. It would also be nice if `find_dynamicsymbols` worked with vectors. A current hack is to convert to a matrix then use these functions.

### PR DESCRIPTION
**Pull request by tosalonijain**
Fixes #13398
I have made changes in the physics module specifically. As the issue demanded that It would be nice to be able to call free_symbols and find_dynamicsymbols on a physics vector. A current hack is to convert to a matrix then use these functions.
So I made arrangements to internally transform a vector into a matrix provided the reference frame as argument and then use the same technique to find free_dynamicsymbols as for a matrix. I also created a small function in the vector class, named free_symbols that takes the vector and reference frame as arguments, converts the vector into matrix and returns the free_symbols out of it.

### CODE REVIEW

**Comment by gxyd** [comment]
Use 'Fixes #13398' (i.e. add a space). Also currently PR is failing code quality tests, you can run code quality tests read this https://github.com/sympy/sympy/wiki/Development-workflow#be-sure-that-all-tests-of-sympy-pass .

File: sympy/physics/mechanics/functions.py [hunk 1]
```diff
@@ -399,7 +399,7 @@ def Lagrangian(frame, *body):
     return kinetic_energy(frame, *body) - potential_energy(*body)
 
 
-def find_dynamicsymbols(expression, exclude=None):
+def find_dynamicsymbols(expression, exclude=None, reference_frame=None):
```
Comment (moorepants) [inline sympy/physics/mechanics/functions.py]: Please include an explanation of this new argument in the docstring for this function. An example of use it helpful too.

File: sympy/physics/vector/vector.py [hunk 2]
```diff
@@ -694,6 +694,8 @@ def applyfunc(self, f):
         for v in self.args:
             d[v[1]] = v[0].applyfunc(f)
         return Vector(d)
+    def free_symbols(self, reference_frame):
```
Comment (moorepants) [inline sympy/physics/vector/vector.py]: need space above

Comment (moorepants) [reply sympy/physics/vector/vector.py]: needs a docstring

Comment (moorepants) [reply sympy/physics/vector/vector.py]: `free_symbols` is a attribute on typical SymPy objects, not a function. Make use of the `@property` decorator to turn this method into a property.

**Comment by moorepants** [comment]
Thanks for the submission. You will need some unit tests for this code.

**Review by moorepants — CHANGES_REQUESTED** [review]

**Comment by tosalonijain** [comment]
I have tried to meet the required changes demanded. Please review.
Thanks

**Comment by tosalonijain** [comment]
@moorepants and @gxyd Please guide me ahead with this issue.
Thanks

File: sympy/physics/mechanics/functions.py [hunk 3]
```diff
@@ -413,6 +413,26 @@ def find_dynamicsymbols(expression, exclude=None):
 
     >>> find_dynamicsymbols(expr, [x, y])
     {Derivative(x(t), t)}
+
+    Optional Parameter
```
[the remark this answers is not in the document]
Comment (tosalonijain) [reply sympy/physics/mechanics/functions.py]: @moorepants I changed it to Parameters. I had a doubt, that in my docstring for find_dynamicsymbols, do I need to specify that when I am considering "expression" to be a matrix and when as a vector.This is so because my code is failing checks with the exception mentioned:


 File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/sympy-1.1.2.dev0-py3.6.egg/sympy/physics/mechanics/functions.py", line 448, in find_dynamicsymbols

    expr=expression.to_matrix(reference_frame)

  File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/sympy-1.1.2.dev0-py3.6.egg/sympy/matrices/matrices.py", line 1852, in __getattr__

    "%s has no attribute %s." % (self.__class__.__name__, attr))

AttributeError: MutableDenseMatrix has no attribute to_matrix.

Although I am intended to employ "to_matrix" on a vector only not on a MutableDenseMatrix.

File: sympy/physics/mechanics/functions.py [hunk 4]
```diff
@@ -413,6 +413,26 @@ def find_dynamicsymbols(expression, exclude=None):
...
+        The frame with respect to which the dynamic symbols of the
+        given vector is to be determined.
+
+    If we intend to apply this function on a vector, the optional
+    ''reference_frame'' is also used to inform about the corresponding frame
+    with respect to which the dynamic symbols of the given vector is to be
+    determined.
+
+    >>> from sympy.physics.mechanics import dynamicsymbols, find_dynamicsymbols, Ref
+    erenceFrame
```
[the remark this answers is not in the document]
Comment (tosalonijain) [reply sympy/physics/mechanics/functions.py]: >>> from sympy.physics.mechanics import (dynamicsymbols, find_dynamicsymbols,
                                         ReferenceFrame)

did you mean to change it to something like this?

[the remark this answers is not in the document]
Comment (gxyd) [reply sympy/physics/mechanics/functions.py]: Use triple backticks \`\`\`  text \`\`\` to format the GitHub comment in a good way.

Anyway, I think some space on second line to make it more in alignment with first line would be good, but I don't think this a strong requirement as your comment also looks just fine.

File: sympy/physics/mechanics/functions.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (tosalonijain) [reply sympy/physics/mechanics/functions.py]: @moorepants and @gxyd please guide over this issue.

File: sympy/physics/mechanics/functions.py [hunk 5]
```diff
@@ -422,8 +441,13 @@ def find_dynamicsymbols(expression, exclude=None):
             raise TypeError("exclude kwarg must be iterable")
     else:
         exclude_set = set()
-    return set([i for i in expression.atoms(AppliedUndef, Derivative) if
-            i.free_symbols == t_set]) - exclude_set
+    if reference_frame==None:
```
Comment (moorepants) [inline sympy/physics/mechanics/functions.py]: `if reference_frame is None:`

File: sympy/physics/mechanics/functions.py [hunk 6]
```diff
@@ -422,8 +441,13 @@ def find_dynamicsymbols(expression, exclude=None):
...
+        return set([i for i in expression.atoms(AppliedUndef, Derivative) if
+                i.free_symbols == t_set]) - exclude_set
+    else:
+        expr=expression.to_matrix(reference_frame)
+        return set([i for i in expr.atoms(AppliedUndef, Derivative) if
+                i.free_symbols == t_set]) - exclude_set
```
Comment (moorepants) [inline sympy/physics/mechanics/functions.py]: I think this might work better if you use duck typing. For example:

```python
try:
    expr = expression.to_matrix(reference_frame)
except AttributeError:  # object doesn't have .to_matrix()
    # code that computes on sympy expression
```

You may need a check to tell the user to add a reference frame if they pass in a vector without one.

Comment (tosalonijain) [reply sympy/physics/mechanics/functions.py]: @moorepants I have made the changes requested. Please consider my modifications and guide accordingly. I am sorry for my passivity towards this issue for the last 10-12 days.

**Comment by tosalonijain** [comment]
@gxyd please guide me ahead to finish this issue

**Comment by gxyd** [comment]
Sorry I may not find time to review this week atleast, may be someone else can review.

File: sympy/physics/mechanics/functions.py [hunk 7]
```diff
@@ -411,8 +412,29 @@ def find_dynamicsymbols(expression, exclude=None):
...
+    ''reference_frame'' is also used to inform about the corresponding frame
+    with respect to which the dynamic symbols of the given vector is to be
+    determined.
+
+    >>> from sympy.physics.mechanics import dynamicsymbols, find_dynamicsymbols
+    >>> from sympy.physics.mechanics import ReferenceFrame
+    >>> a, b, c = dynamicsymbols('a, b, c')
+    >>> A = ReferenceFrame('A')
+    >>> v = a * A.x + b * A.y + c * A.z
+    >>> find_dynamicsymbols(v,reference_frame=A)
```
Comment (moorepants) [inline sympy/physics/mechanics/functions.py]: space after comma, see PEP8

File: sympy/physics/mechanics/functions.py [hunk 8]
```diff
@@ -422,8 +444,17 @@ def find_dynamicsymbols(expression, exclude=None):
...
     else:
         exclude_set = set()
-    return set([i for i in expression.atoms(AppliedUndef, Derivative) if
-            i.free_symbols == t_set]) - exclude_set
+    try:
+        if isinstance(expression, Vector):
+            expr=expression.to_matrix(reference_frame)
+            return set([i for i in expr.atoms(AppliedUndef, Derivative) if
+                    i.free_symbols == t_set]) - exclude_set
+    except TypeError:
```
Comment (moorepants) [inline sympy/physics/mechanics/functions.py]: I'm not sure this is the best check here. Why not check explicitly? Like "if isinstance(expression, Vector) and reference_frame=None".

File: sympy/physics/mechanics/functions.py [hunk 9]
```diff
@@ -422,8 +444,17 @@ def find_dynamicsymbols(expression, exclude=None):
...
+        print(" function missing one required argument of type ReferenceFrame ")
```
Comment (moorepants) [inline sympy/physics/mechanics/functions.py]: Make a complete sentence here that is more informative to the user. "You must provide a reference frame when passing in a vector expression."

File: sympy/physics/mechanics/functions.py [hunk 10]
```diff
@@ -422,8 +444,17 @@ def find_dynamicsymbols(expression, exclude=None):
...
+        return
+    if reference_frame is None :
+        return set([i for i in expression.atoms(AppliedUndef, Derivative) if
+                i.free_symbols == t_set]) - exclude_set
```
Comment (moorepants) [inline sympy/physics/mechanics/functions.py]: This line is repeated and should be pulled out of the type checking code.

File: sympy/physics/mechanics/tests/test_functions.py [hunk 11]
```diff
@@ -165,6 +166,12 @@ def test_find_dynamicsymbols():
     sol = {x, y.diff(), y, x.diff().diff(), z, z.diff()}
     assert find_dynamicsymbols(expr) == sol
     # Test finding all but those in sym_list
-    exclude = [x, y, z]
+    exclude_list = [x, y, z]
     sol = {y.diff(), x.diff().diff(), z.diff()}
-    assert find_dynamicsymbols(expr, exclude) == sol
+    assert find_dynamicsymbols(expr, exclude=exclude_list) == sol
```
Comment (moorepants) [inline sympy/physics/mechanics/tests/test_functions.py]: It is best not to change the API if at all possible. Why is this change required?

File: sympy/physics/vector/vector.py [hunk 12]
```diff
@@ -695,6 +695,19 @@ def applyfunc(self, f):
             d[v[1]] = v[0].applyfunc(f)
         return Vector(d)
 
+    def free_symbols(self, reference_frame):
+        """
+        Returns from the vector those which are free symbols.
```
Comment (moorepants) [inline sympy/physics/vector/vector.py]: "Returns the free symbols in the measure numbers of the vector expressed in the given reference frame."

File: sympy/physics/vector/vector.py [hunk 13]
```diff
@@ -695,6 +695,19 @@ def applyfunc(self, f):
...
+
+        Parameter
+        ---------
```
Comment (moorepants) [inline sympy/physics/vector/vector.py]: Be consistent. Use either `----` or `===`.

File: sympy/physics/vector/vector.py [hunk 14]
```diff
@@ -695,6 +695,19 @@ def applyfunc(self, f):
...
+        reference_frame : ReferenceFrame
+            The frame with respect to which the dynamic symbols of the
```
Comment (moorepants) [inline sympy/physics/vector/vector.py]: This function does not find dynamic symbols.

**Review by moorepants — CHANGES_REQUESTED** [review]
Please address the comments I've left. Once done, this should be ready to merge. Thanks!

File: sympy/physics/mechanics/functions.py [hunk 15]
```diff
@@ -411,8 +412,29 @@ def find_dynamicsymbols(expression, exclude=None):
     If the optional ``exclude`` kwarg is used, only dynamicsymbols
     not in the iterable ``exclude`` are returned.
 
-    >>> find_dynamicsymbols(expr, [x, y])
+    >>> find_dynamicsymbols(expr, exclude=[x, y])
```
[the remark this answers is not in the document]
Comment (tosalonijain) [reply sympy/physics/mechanics/functions.py]: SInce there are two optional parameters now, "exclude" and "reference_frame", unlike before where there was only one i.e. "exclude".

File: sympy/physics/mechanics/functions.py [hunk 16]
```diff
@@ -411,8 +412,29 @@ def find_dynamicsymbols(expression, exclude=None):
...
     {Derivative(x(t), t)}
+
+    Parameters
+    ==========
```
[the remark this answers is not in the document]
Comment (tosalonijain) [reply sympy/physics/mechanics/functions.py]: @moorepants can you please tell me what are the class types for the other two parameter, i.e. exclude and expression?
Is it okay with this:
Parameters
    ==========

    expression : ImmutableDenseMatrix or Vector

    reference_frame : ReferenceFrame
        The frame with respect to which the dynamic symbols of the
        given vector is to be determined.

    exclude : kwarg constituting of the dynamicsymbols to be excluded.

Though I know expression covers a broader range of classes, So please specify what should I be mentioning here..Thanks!

File: sympy/physics/mechanics/functions.py (hunk 10 above)
Comment (tosalonijain) [reply sympy/physics/mechanics/functions.py]: I have changed it to:

if isinstance(expression, Vector) and reference_frame is None :
            raise TypeError("You must provide a reference frame when passing in a vector expression.")
            return
    if isinstance(expression, Vector) :
        expression=expression.to_matrix(reference_frame)
    return set([i for i in expression.atoms(AppliedUndef, Derivative) if
            i.free_symbols == t_set]) - exclude_set

without using any try exception blocks, does it seem fine?

File: sympy/physics/vector/vector.py (hunk 12 above)
Comment (tosalonijain) [reply sympy/physics/vector/vector.py]: @moorepants apologies for such silly mistakes. I have revised the function free_symbols, please check:

    def free_symbols(self, reference_frame):
        """
        Returns the free symbols in the measure numbers of the vector
        expressed in the given reference frame.

        Parameter
        =========

        reference_frame : ReferenceFrame
            The frame with respect to which the free symbols of the
            given vector is to be determined.

        """

        return self.to_matrix(reference_frame).free_symbols

File: sympy/physics/mechanics/functions.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (moorepants) [reply sympy/physics/mechanics/functions.py]: Got it. I see now that exclude was a previous optional parameter. This change is fine.

File: sympy/physics/mechanics/functions.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (moorepants) [reply sympy/physics/mechanics/functions.py]: exclude is a sequence of dynamic symbols

expression is a sympy expression

[the remark this answers is not in the document]
Comment (moorepants) [reply sympy/physics/mechanics/functions.py]: or "iterable of dynamic symbols" for the exclude

File: sympy/physics/mechanics/functions.py (hunk 10 above)
Comment (moorepants) [reply sympy/physics/mechanics/functions.py]: I like this better. Write unit tests to test whether it is fine. You don't need the first return statement.

File: sympy/physics/mechanics/tests/test_functions.py (hunk 11 above)
Comment (moorepants) [reply sympy/physics/mechanics/tests/test_functions.py]: ignore this comment

File: sympy/physics/vector/vector.py (hunk 12 above)
Comment (moorepants) [reply sympy/physics/vector/vector.py]: Looks good. These are not mistakes or silly. I'm just showing you some of the conventions we use in SymPy to keep the code and doc quality high.

File: sympy/physics/mechanics/functions.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (tosalonijain) [reply sympy/physics/mechanics/functions.py]: is this fine??

    Parameters
    ==========

    expression : sympy expression

    reference_frame : ReferenceFrame
        The frame with respect to which the dynamic symbols of the
        given vector is to be determined.

    exclude : iterable of dynamicsymbols

File: sympy/physics/mechanics/tests/test_functions.py (hunk 11 above)
Comment (tosalonijain) [reply sympy/physics/mechanics/tests/test_functions.py]: @moorepants I have made a new commit addressing the requested changes. Please review.
Thanks.

**Comment by tosalonijain** [comment]
@moorepants if possible please help me completing this issue by reviewing the pull request.

**Comment by shivamvats** [comment]
@tosalonijain A few of the changes requested by @moorepants are yet to addressed (see https://github.com/sympy/sympy/pull/13549/files/1a9306301c6361af05f5b218ce1b3a2973569e3a).

**Comment by tosalonijain** [comment]
@shivamvats sorry but I am unable to spot the unaddressed changes. It would be great if you could help me with that.
Thanks.

**Comment by tosalonijain** [comment]
@gxyd please review this and help me finish this issue.

**Comment by gxyd** [comment]
The link left by Shivam was probably broken (for whatever reason). Here's the unaddressed comment link https://github.com/sympy/sympy/pull/13549#discussion_r147588615 . Let us know if you still have any trouble finding the comment.

**Comment by tosalonijain** [comment]
@gxyd I hope you are asking me consider this:
"Please include an explanation of this new argument in the docstring for this function. An example of use it helpful too."

For this I had done the following : 

"""Find all dynamicsymbols in expression.

    If the optional ``exclude`` kwarg is used, only dynamicsymbols
    not in the iterable ``exclude`` are returned.
    If we intend to apply this function on a vector, the optional
    ''reference_frame'' is also used to inform about the corresponding frame
    with respect to which the dynamic symbols of the given vector is to be
    determined.

    Parameters
    ==========

    expression : sympy expression

    reference_frame : ReferenceFrame
        The frame with respect to which the dynamic symbols of the
        given vector is to be determined.

    exclude : iterable of dynamicsymbols

    Examples
    ========

    >>> from sympy.physics.mechanics import dynamicsymbols, find_dynamicsymbols
    >>> from sympy.physics.mechanics import ReferenceFrame
    >>> x, y = dynamicsymbols('x, y')
    >>> expr = x + x.diff()*y
    >>> find_dynamicsymbols(expr)
    {x(t), y(t), Derivative(x(t), t)}
    >>> find_dynamicsymbols(expr, exclude=[x, y])
    {Derivative(x(t), t)}
    >>> a, b, c = dynamicsymbols('a, b, c')
    >>> A = ReferenceFrame('A')
    >>> v = a * A.x + b * A.y + c * A.z
    >>> find_dynamicsymbols(v, reference_frame=A)
    {a(t), b(t), c(t)}

    """

**Comment by moorepants** [comment]
This is quite close to being merged. Was it accidentally closed?

**Comment by tosalonijain** [comment]
@moorepants yes I guess it was accidentally closed. Guide me what to be done to get it merged.
Thanks

File: sympy/physics/mechanics/functions.py [hunk 17]
```diff
@@ -399,20 +400,44 @@ def Lagrangian(frame, *body):
...
+
+    expression : sympy expression
+
+    reference_frame : ReferenceFrame
+        The frame with respect to which the dynamic symbols of the
+        given vector is to be determined.
+
+    exclude : iterable of dynamicsymbols
```
Comment (gxyd) [inline sympy/physics/mechanics/functions.py]: Can you please document them in the order in which they appear in the API? i.e. first `expression`, then `exclude` and lastly `reference_frame`.

Also use this format for `exclude` and `reference_frame`: 
```
exclude : iterable of dynamicsymbols, optional

reference_frame : ReferenceFrame, optional
    The frame with respect to which the dynamic symbols of the
    given vector is to be determined.
```

File: sympy/physics/mechanics/functions.py [hunk 18]
```diff
@@ -422,6 +447,10 @@ def find_dynamicsymbols(expression, exclude=None):
             raise TypeError("exclude kwarg must be iterable")
     else:
         exclude_set = set()
+    if isinstance(expression, Vector) and reference_frame is None :
+            raise TypeError("You must provide a reference frame when passing in a vector expression.")
```
Comment (gxyd) [inline sympy/physics/mechanics/functions.py]: rather I would say, raise a `ValueError` error here not `TypeError`, see: https://docs.python.org/2/library/exceptions.html#exceptions.ValueError

Here since it is expected that user passes non-None `reference_frame`. Also use a better error message like:
`You must provide reference_frame when passing a vector expression, got %s." % reference_frame`

File: sympy/physics/mechanics/functions.py [hunk 19]
```diff
@@ -422,6 +447,10 @@ def find_dynamicsymbols(expression, exclude=None):
...
+    if isinstance(expression, Vector) :
+        expression=expression.to_matrix(reference_frame)
```
Comment (gxyd) [inline sympy/physics/mechanics/functions.py]: add a space here `expression = expression.to_matrix(reference_frame)`

File: sympy/physics/mechanics/functions.py (hunk 18 above)
Comment (gxyd) [reply sympy/physics/mechanics/functions.py]: Also I would write this as:
```
if isinstance(expression, Vector):
    if reference_frame is None:
        raise ValueError("You must provide reference_frame when passing a "
                         "vector expression, got %s." % reference_frame)
    else:
        expression = expression.to_matrix(reference_frame)
```

Also make sure character length on line is <= 79.

**Review by gxyd — COMMENTED** [review]
Apart from that LGTM.

**Comment by gxyd** [comment]
Also sorry, I think the comment by Jason I was referring to has been addressed. I just left a few comments, apart from that it looks good to me.

File: sympy/physics/mechanics/functions.py [hunk 20]
```diff
@@ -422,6 +447,11 @@ def find_dynamicsymbols(expression, exclude=None):
             raise TypeError("exclude kwarg must be iterable")
     else:
         exclude_set = set()
+    if isinstance(expression, Vector) and reference_frame is None :
```
Comment (gxyd) [inline sympy/physics/mechanics/functions.py]: change this to:
```python
if isinstance(expresssion, Vector):
    if reference_frame is None:
        raise ValueError(" ... ")
    else:
        expression = expression.to_matrix(reference_frame)
```

Also I would expect a test case for this as well.

File: sympy/physics/mechanics/functions.py (hunk 19 above)
Comment (tosalonijain) [reply sympy/physics/mechanics/functions.py]: I don't understand why the build is not succeeding. It is showing the following error:

File "/home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/sympy-1.1.2.dev0-py2.7.egg/sympy/combinatorics/coset_table.py", line 910, in sympy.combinatorics.coset_table.coset_enumeration_r

Failed example:

    index

Expected:

    500

Got:

    1 

Although I haven't even touched these files.

Comment (jksuom) [reply sympy/physics/mechanics/functions.py]: It seems that the coset table algorithm has some random methods that may sometimes lead to problems. That has nothing to do with your PR. The failing job will most probably pass when restarted.

File: sympy/physics/mechanics/functions.py (hunk 20 above)
Comment (tosalonijain) [reply sympy/physics/mechanics/functions.py]: I have made the requested change. A test case for this is already present as:

 # Test finding all dynamicsymbols in a vector with a given reference frame
    d, e, f = dynamicsymbols('d, e, f')
    A = ReferenceFrame('A')
    v = d * A.x + e * A.y + f * A.z
    sol = {d, e, f}
    assert find_dynamicsymbols(v, reference_frame=A) == sol

File: sympy/physics/mechanics/functions.py [hunk 21]
```diff
@@ -11,6 +11,7 @@
 from sympy import simplify
 from sympy.core.backend import (Matrix, sympify, Mul, Derivative, sin, cos,
                                 tan, AppliedUndef, S)
+from sympy.matrices.immutable import ImmutableDenseMatrix
```
Comment (gxyd) [inline sympy/physics/mechanics/functions.py]: Is this import used anywhere?

File: sympy/physics/mechanics/functions.py [hunk 22]
```diff
@@ -422,6 +447,12 @@ def find_dynamicsymbols(expression, exclude=None):
             raise TypeError("exclude kwarg must be iterable")
     else:
         exclude_set = set()
+    if isinstance(expression, Vector):
+        if reference_frame is None:
+            raise ValueError("You must provide reference_frame when passing a "
```
Comment (gxyd) [inline sympy/physics/mechanics/functions.py]: I mean adding a test case for this `raise ValueError`, i.e. add test case which verifies that a `ValueError` is indeed raised when this condition is satisfied.

File: sympy/physics/mechanics/functions.py (hunk 21 above)
Comment (tosalonijain) [reply sympy/physics/mechanics/functions.py]: No, its not being used anywhere. I have removed it.

File: sympy/physics/mechanics/functions.py (hunk 22 above)
Comment (tosalonijain) [reply sympy/physics/mechanics/functions.py]: @gxyd I could not find how to write a test case for checking if a ValueError is raised or not. Will it be done using assertRaises ?
Please guide. Thanks

Comment (gxyd) [reply sympy/physics/mechanics/functions.py]: See this for example: https://github.com/sympy/sympy/blob/master/sympy/functions/combinatorial/tests/test_comb_numbers.py#L88

Comment (tosalonijain) [reply sympy/physics/mechanics/functions.py]: Its showing something like this:

>>> from sympy.physics.mechanics.tests.test_functions import test_find_dynamicsymbols
>>> test_find_dynamicsymbols()Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/saloni/sympy/sympy/physics/mechanics/tests/test_functions.py", line 181, in test_find_dynamicsymbols
    raises(ValueError, lambda: find_dynamicsymbols(v))
NameError: name 'raises' is not defined

Comment (gxyd) [reply sympy/physics/mechanics/functions.py]: Ofcourse you will need to import `raises`, its a sympy utility, not python's default.

Comment (tosalonijain) [reply sympy/physics/mechanics/functions.py]: @gxyd @moorepants I guess I have addressed all the requested changes. If not do let me know. I also couldn't figure out why the build is failing. Looking forward for it to be merged.
Thanks

File: sympy/physics/mechanics/functions.py [hunk 23]
```diff
@@ -422,6 +446,12 @@ def find_dynamicsymbols(expression, exclude=None):
...
+                             "vector expression, got %s." % reference_frame")
```
Comment (jksuom) [inline sympy/physics/mechanics/functions.py]: `reference_frame` should probably not end with ".

Comment (tosalonijain) [reply sympy/physics/mechanics/functions.py]: yes, that was my mistake. I have rectified it.

**Review by gxyd — APPROVED** [review]
LGTM

**Comment by gxyd** [comment]
@tosalonijain also please make sure that after this PR is merged, we get this listed in Release notes in minor changes section (?) https://github.com/sympy/sympy/wiki/Release-Notes-for-1.2#minor-changes

**Comment by gxyd** [comment]
@moorepants I think your review comment have been addressed, so I'm merging this.

**Comment by gxyd** [comment]
Thanks @tosalonijain . Can you please update the release notes?

**Comment by gxyd** [comment]
Ping @tosalonijain will you have the time to update release notes for this change?

Ah, this I guess one big benifit we will be having with release note with the PR itself :)

**Comment by tosalonijain** [comment]
@gxyd I am really sorry for my passivity. Instruct me please what updates to be made and where to be made.
Thanks.

**Comment by tosalonijain** [comment]
@gxyd I have updated the release notes. Please check if it seems okay to you and suggest changes if not.
Thanks!

**Comment by gxyd** [comment]
Looks good. Thanks for getting back on that.
